File: backend/app/schemas/publishing.py
"""Pydantic v2 schemas for Module 8 (Publishing / Distribution).

Token material (access/refresh tokens) is never exposed through any response
schema in this module.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.models.publishing import (
    PublishDestinationType,
    PublishJobStatus,
    SocialAccountStatus,
    SocialPlatform,
)
from app.public_url import public_base_url


class SocialAccountResponse(BaseModel):
    """A connected destination. Deliberately omits every *_token_encrypted field."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    platform: SocialPlatform
    external_account_id: str | None = None
    display_name: str | None = None
    status: SocialAccountStatus
    token_expires_at: datetime | None = None
    created_at: datetime | None = None


class ConnectStartResponse(BaseModel):
    """Returned by POST /social-accounts/connect/{platform}."""

    auth_url: str
    state: str


class PublishRequest(BaseModel):
    """Body of POST /clips/{id}/publish.

    ``scheduled_at`` null (or in the past) => publish immediately.
    ``slack_webhook_url`` is required for destination_type slack/teams and is
    stored encrypted on a SocialAccount row; it is never returned.
    """

    destination_type: PublishDestinationType
    social_account_id: int | None = None
    slack_webhook_url: str | None = None
    caption_text: str | None = Field(default=None, max_length=5000)
    scheduled_at: datetime | None = None


class PublishJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    clip_id: int
    user_id: int
    social_account_id: int | None = None
    destination_type: PublishDestinationType
    caption_text: str | None = None
    scheduled_at: datetime | None = None
    published_at: datetime | None = None
    status: PublishJobStatus
    external_post_id: str | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def share_url(self) -> str | None:
        """Public URL for a completed `link` job (external_post_id is the slug)."""
        if (
            self.destination_type == PublishDestinationType.link
            and self.external_post_id
        ):
            return f"{public_base_url()}/s/{self.external_post_id}"
        return None


class PublishJobUpdate(BaseModel):
    """Reschedule / edit a job that is still draft or scheduled."""

    scheduled_at: datetime | None = None
    caption_text: str | None = Field(default=None, max_length=5000)


class CalendarEntry(BaseModel):
    """One chip on the publish calendar."""

    id: int
    clip_id: int
    title: str | None = None
    platform: str
    scheduled_at: datetime | None = None
    status: PublishJobStatus


class ShareLinkCreate(BaseModel):
    expires_at: datetime | None = None


class ShareLinkResponse(BaseModel):
    url: str
    slug: str
    is_active: bool
    expires_at: datetime | None = None
    view_count: int


class PublicClipResponse(BaseModel):
    """Unauthenticated payload for GET /s/{slug}."""

    title: str | None = None
    video_url: str
    duration: float

File: frontend/src/types/publishing.ts
// Module 8 (Publishing / Distribution) — request/response shapes not covered by
// src/types/index.ts. index.ts is not edited; module-local types live here.

import type {
  PublishDestinationType,
  PublishJob as BasePublishJob,
  PublishJobStatus,
  ShareLink,
  SocialAccount,
  SocialAccountStatus,
  SocialPlatform,
} from './index';

/* -------------------------------------------------------------------------- */
/* Re-exports + aliases                                                        */
/* -------------------------------------------------------------------------- */

export type { ShareLink, SocialAccount, SocialAccountStatus, PublishJobStatus };

/** Social / distribution platform. */
export type Platform = SocialPlatform;

/** Where a clip gets published. */
export type DestinationType = PublishDestinationType;

/** Platforms backed by a real OAuth social account. */
export const SOCIAL_PLATFORMS: readonly Platform[] = [
  'tiktok',
  'instagram',
  'youtube',
] as const;

/** Every platform we can connect (social + chat). */
export const CONNECTABLE_PLATFORMS: readonly Platform[] = [
  'tiktok',
  'instagram',
  'youtube',
  'slack',
  'teams',
] as const;

/** All publish-job statuses, in lifecycle order. */
export const PUBLISH_JOB_STATUSES: readonly PublishJobStatus[] = [
  'draft',
  'scheduled',
  'publishing',
  'published',
  'failed',
] as const;

export const PLATFORM_LABELS: Record<Platform, string> = {
  tiktok: 'TikTok',
  instagram: 'Instagram',
  youtube: 'YouTube',
  slack: 'Slack',
  teams: 'Microsoft Teams',
};

/* -------------------------------------------------------------------------- */
/* Publish jobs                                                                */
/* -------------------------------------------------------------------------- */

/**
 * A publish job as returned by the list / calendar endpoints. The backend
 * denormalises the clip title and resolved platform onto the row for display.
 */
export interface PublishJob extends BasePublishJob {
  clip_title?: string | null;
  platform?: Platform | null;
}

/** GET /publish-jobs query filters. */
export interface PublishJobFilters {
  status?: PublishJobStatus;
  from?: string;
  to?: string;
}

/** PUT /publish-jobs/{id} — reschedule / edit a scheduled job. */
export interface UpdatePublishJobInput {
  scheduled_at?: string | null;
  caption_text?: string | null;
}

/* -------------------------------------------------------------------------- */
/* Publish action (POST /clips/{id}/publish)                                   */
/* -------------------------------------------------------------------------- */

interface PublishInputBase {
  caption_text: string;
  scheduled_at?: string | null;
}

export interface SocialPublishInput extends PublishInputBase {
  destination_type: 'social';
  social_account_id: number;
}

export interface WebhookPublishInput extends PublishInputBase {
  destination_type: 'slack' | 'teams';
  slack_webhook_url: string;
}

export interface LinkPublishInput extends PublishInputBase {
  destination_type: 'link';
}

/** Discriminated union accepted by POST /clips/{id}/publish. */
export type PublishInput =
  | SocialPublishInput
  | WebhookPublishInput
  | LinkPublishInput;

/** Draft state held by <DestinationPicker> before it is narrowed to PublishInput. */
export interface DestinationDraft {
  destination_type: DestinationType;
  social_account_id: number | null;
  webhook_url: string;
}

export const EMPTY_DESTINATION_DRAFT: DestinationDraft = {
  destination_type: 'social',
  social_account_id: null,
  webhook_url: '',
};

/* -------------------------------------------------------------------------- */
/* Connect flow                                                                */
/* -------------------------------------------------------------------------- */

/** POST /social-accounts/connect/{platform}. */
export interface ConnectAccountResponse {
  auth_url: string;
  state: string;
}

/* -------------------------------------------------------------------------- */
/* Calendar                                                                    */
/* -------------------------------------------------------------------------- */

/** GET /publish/calendar?from=&to= — one scheduled entry. */
export interface CalendarEntry {
  publish_job_id: number;
  clip_id: number;
  clip_title: string;
  destination_type: DestinationType;
  platform: Platform | null;
  scheduled_at: string;
  status: PublishJobStatus;
  caption_text: string | null;
}

/** Inclusive ISO date range for calendar / job queries. */
export interface DateRange {
  from: string;
  to: string;
}

/* -------------------------------------------------------------------------- */
/* Share links                                                                 */
/* -------------------------------------------------------------------------- */

/** POST /clips/{id}/share-link. */
export interface ShareLinkResponse {
  url: string;
  slug: string;
  is_active: boolean;
  expires_at: string | null;
  view_count: number;
}

/* -------------------------------------------------------------------------- */
/* Public clip (GET /s/{slug}) — unauthenticated                               */
/* -------------------------------------------------------------------------- */

export interface PublicClip {
  title: string;
  video_url: string;
  duration: number;
}
